[cooper] 미션3 : 체스판 초기화

pbg0205 opened the pull request:
안녕하세요.

코드스쿼드 2021 마스터즈 백엔드 과정 Cooper 입니다.
미션3 : 체스판 초기화 리뷰 부탁드립니다.

이전 피드백에서 why에 관한 내용이 많이 부족한거 같아 아래 추가적으로 변경사항에 대한 고민을 적어보았습니다. 아직 많이 부족하지만 추가적으로 수정할 부분이나 의견있으시면 거침없이 패드백 부탁드립니다! 별도로 시간내어 리뷰해주시는 점 미리 감사하다는 말씀드립니다!
잘 부탁드립니다!

### 미션3 변경사항
-  Pawn 클래스 멤버 변수 representaion을 추가하였습니다.
Pawn의 represetation 추가 방법에 대해 고민하였습니다. color와 representation을 매개인자로 받는 생성자를 선언하는 방법도 고민해보았지만 Color enum에 내부 변수를 별도로 선언해 representation을 초기화하면 별도의 예외 처리가 필요하지 않을 것 같다고 판단했습니다. representation enum을 선언하는 방법도 고려해보았습니다만 폰 하나만 존재하는 상황에서 색상에 따라서 representation을 할당할 수 있기 때문에 representation enum은 잠시 보류해두었고 우선적으로 내부 변수로 할하는 방법으로 구현 진행하였습니다.

- Board 클래스의 리스트를 HashMap으로 관리하도록 변경하였습니다.
 흰색과 검정색 폰을 관리하기 위해 whitePawnList와 blackPawnList를 별도로 관리하였습니다. 하지만 코드를 작성하는 과정에서 같은 기능이지만 다른 List를 사용하기 위한 중복 코드가 계속해서 발생하였습니다.
 각각의 color에 따라서 리스트를 참조할 수 있다면 중복되는 메서드가 줄어들고 리스트를 관리하기에 더욱 용이하다고 판단했습니다. 그래서 Color를 Key값, List<Pawn>를  value로 가지는 HashMap를 선언하여 색상에 따른 폰리스트를 관리할 수 있도록 구현했습니다.

- Board print 메서드를 구현했습니다.
 폰들은 배치되는 각각의 행이 정해져 있어 해당 행일 경우에만 List에 존재하는 폰들을 출력하는 방식으로 구현 진행하였습니다. 그래서 Board 클래스에 폰이 초기화되는 행 번호를 상수로 선언하였고 각각의 폰이 위치하는 행을 추가하였습니다.

- 메인클래스 로직을 추가하였습니다.
 while문에 조건으로 isPlaying이라는 boolean 변수를 이용해 현재 게임 중인 상태를 확인하고 start 명령어를 입력받을 경우 현재 초기화된 상태를 출력하고 end 명령어를 입력받을 경우, isPlaying = false 형태로 변경하여 게임이 종료되도록 구현하였습니다.

pbg0205 committed:
test : BoardTest 인덱스외 범위 발생 예외 변경

- findPawn 메서드의 잘못된 인덱스 매개인자를 입력받았을 때 발생하는
예외로 변경하는 것이 적합하다고 판단하여  IllegalArgumentException으로
예외를 변경했습니다.

pbg0205 committed:
refactor : Board 클래스 findPawn 예외 종류 변경

- findPawn 메서드는 매개인자 index를 매개인자로 받아 해당 인덱스를
탐색하는 메서드이므로 IllegalArgumentException을 사용하는 것이 적합하
다고 판단했습니다.

pbg0205 committed:
style : build.gradle 개행 추가 설정

pbg0205 committed:
feat : Color 에넘 상수 내 representation

- 색상에 따라 representation이 각각 정해져 있기 때문에
 represenation을생성자의 매개인자로 추가하는 것보다 Color를 매개인자로
받은 후, color에서 representation을 할당하도록 구현했습니다.

pbg0205 committed:
feat : Pawn 클래스 멤버 변수 representation 추가

- Pawn의 상태를 표현하는 represetaion을 멤버 변수로 추가했습니다.

- Color 에넘에 representation을 할당할 경우,  각 색상에 맞는 값을
할당할 수 있는 점에서  하나의 매개변수 사용하는 방식으로 구현했습니다.

pbg0205 committed:
test : PawnTest 클래스 create 메서드 추가

- verifyPawn 메서드를 추가햇습니다. 내부는 이전에 피드백을 받았던
AssertAll을 사용했습니다. AssertAll을 사용할 경우, 테스트가 피해해도
테스트를 멈추지 않고 모든 테스트를 확인할 수 있어 각각의 테스트 코드를
 확인할 때 용이하다고 판단했습니다.

pbg0205 committed:
feat : Board 클래스 자료구조 변경 밑 출력 기능 추가

- 각각의 색상에 따른 리스트를 관리하기 위해 Color를 key를 가지는 HashMap
으로 변경하였습니다. 각각의 List를 사용할 경우, 같은 기능의 메서드를
각각 리스트에 맞게 메서드 작성이 필요해  중복 코드가 방생하였습니다.

-  반면, Map을 사용할 경우, key 값을 통해서 리스트를 참조하기 때문에
중복 코드를 작성하지 않고 메서드를 작성할 수 있는 장점있어  데이터 관리
및  중복 코드 제거 면에서 적합하다고 판단했습니다.

- print 메서드를  선언해 각 행의 결과를 출력할 수 있습니다. 리스트를 순회
하면서 representation을 추가하기 때문에 StringBuilder를 사용하는 것이
적합하다고 판단했고 각각의 결과는 요구사항에 맞게  String으로 반환하였
습니다.

- initialize 메서드를 작성해 초기 폰 상태를 초기화 하였습니다.

pbg0205 committed:
test : BoardTest 클래스 테스트 추가

- initialize 메서드의 테스트 코드를 assertAll을 이용해서 테스트 정상
작동 여부를 확인 하였습니다.

- print 메서드 정상 출력을 확인하기 위해 테스트 코드를 작성해
정상 작동 여부를  확인 하였습니다.

pbg0205 committed:
feat : Main method 작성

- 시작과 종료에 대한 표준 입력을 받기 위해 Scanner를 사용하였습니다.
- 시작 명령을 받을 경우, board의 상태를 전체 출력할 수 있습니다.
- 종료 명령을 확인할 경우, 프로그램이 정상 종료 됩니다.

imjinbro commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -1,34 +1,101 @@
-}
\ No newline at end of file
+
+    public String print () {
```
print 라는 메소드명으로 되어있어서 판을 출력하는 역할까지 할 것으로 보였는데, 코드 상에서는 체스판만 리턴하고 있네요
Board 클래스 역할을 잘 인지하셔서 프린트 하는 역할(System.out.println)까지 맡지 않도록 한 것까지는 좋았으나 메소드명은 체스판을 가져온다는 의미로 지어주시면 좋겠습니다.

imjinbro replied:
요구사항에 print()를 만들어라고 되어있지만 System.out.println() 이라는 실제 출력하는 역할이 Main 클래스에 있다면 
Board 클래스의 메소드는 출력 역할이 아니므로 코드 읽는 이 또는 협업하는 팀원이 혼동되지않도록 이름을 지어주는 것도 중요합니다~!

pbg0205 replied:
아무래도 print라는 내용이 출력을 의미하는 부분인데 요구사항에 맞추려보니 그대로 명명하게 되었습니다ㅠ 메서드명을 getBoardStatus와 각 행을 탐색하는 메서드명도 printEacchRow -> getEachRowStatus 재명명하였는데 getter 느낌이라서 정확한 명명인지는 확신이 잘 안서네요ㅠ 확인 부탁드립니다!

imjinbro replied:
print 명명보다는 더 나아보여서 좋네요
정답은 없지만 다른 사람이 코드를 읽었을 때 어떨까 라는 생각도 같이 해보시면서 짜시면 조금이라도 도움 되지 않을까 생각해봅니다!
수고하셨습니다

imjinbro commented on src/main/java/cooper/chess/Main.java:
```diff
@@ -1,7 +1,35 @@
+import java.util.Scanner;
+
 public class Main {
     public static void main(String[] args) {
```
지금은 코드가 몇줄 안되어서 읽기 어려운건 아니지만 조금 더 넘어가면 하나라도 놓치면 읽기가 힘들 수 있을듯해요
커맨드에 대한 처리 메소드를 만드신 것처럼 메소드로 묶어보는건 어떨까요?

잡지의 목차(아래)처럼 코드 작성하신다면 한 눈에 코드가 어떤 기능들을 할지 보여서 더 좋을 것 같아요!
1. XX
2. OO
3. TT

imjinbro commented on src/main/java/cooper/chess/Main.java:
```diff
@@ -1,7 +1,35 @@
+
 public class Main {
     public static void main(String[] args) {
+        Scanner scanner = new Scanner(System.in);
```
지금은 코드가 몇줄 안되어서 읽기 어려운건 아니지만 조금 더 넘어가면 하나라도 놓치면 읽기가 힘들 수 있을듯해요
커맨드에 대한 처리 메소드를 만드신 것처럼 메소드로 묶어보는건 어떨까요?

잡지의 목차(아래는 예시입니다)처럼 코드 작성 하신다면 한 눈에 코드가 어떤 기능들을 할지 보여서 더 좋을 것 같아요!

1. 체스판 세팅
2. 체스 실행

pbg0205 replied:
조언 감사합니다!😄 Main class에 체스 게임 내부 로직을 구현하는 것보다 전체적인 게임 로직을 담당하는 클래스에서 별도로 관리하는 것이 나을 것 같아 별도의 ChessGame 클래스를 선언해보았습니다. ChessGame 내부에서 리뷰해주신 내용대로 메서드를 분리시키는 방식으로 진행해보았는데 혹시 부족한 부분 있으면 말씀 부탁드립니다!

imjinbro replied:
메소드로 구분해주니 코드 읽기가 한결 편해졌군요
Main, ChessGame을 굳이 구분하지않고, Main을 ChessGame으로 변경하는 것도 좋은 방법이라고 생각합니다.

imjinbro commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -1,34 +1,101 @@
+    private static final int WHITE_INIT_INDEX = 1;
+    private static final int BLACK_INIT_INDEX = 6;
+
+    private final Map<Color, List<Pawn>> pawnListMap;
```
데이터 타입이 조금 복잡해지는게 느껴지실까요?
컬러가 소유한 pawn을 다루기가 복잡해지실거예요 
하나의 클래스로 분리해보는 것도 추천드립니다.
관련된 링크 하나 남겨두도록 하겠습니다.
https://jojoldu.tistory.com/412

pbg0205 replied:
말씀해주신대로 일급 객체 형태로 구현해보려고 별도로 PawnGroup으로 클래스를 분리해서 작성해보았습니다. 하나 고민이 들었던 부분이 생성자의 매개인자로 Map<Color, List>을 아규먼트로 받아올 때, List 내부에 모든 element들을 추가한 상태도 전달 하는 방법이 불변성을 유지하는 방법인지에 대해 고민했습니다ㅠ Map에 final을 선언해서 Map의 불변성이 유지할 수 있지 않을까 해서 빈 ArrayList를 전달받고 List 안에 element를 추가할 수 있는 방법으로 구현해보았는데 이렇게 구현해도 불변성이 유지가 되는 방법이 맞는지 궁금합니다ㅠ

imjinbro replied:
final을 굳이 지정해주지 않더라도 setter 메소드를 외부에 공개하지않는다면 불변성은 유지가 되겠죠?
저 글에서 언급한 불변성도 위의 이야기로 보입니다.
현재는 Public으로 할당된 메소드가 없다보니 타 플레이어가 뭔가 변경할 수가 없겠네요
객체의 상태(멤버필드 할당된 값)는 객체(본인)가 판단하여 변경한다에 대해서 공부해보시면 좋을 것 같네요

그리고 제가 컬렉션을 클래스 랩핑하길 추천드린건 복잡한 콜렉션 타입을 필드로 가지고 있다보니 
코드가 점점 복잡해지는 면이 있어보여서 추천드리긴 했습니다.
이 방법(랩핑 클래스 만들기) 말고도 다른 방법이 또 있을 수 있으니 이제 변경은 하지 않더라도 앞으로 코딩 하실 때 여러 안에 대해서 고려 해보는 습관을 길러보는 것도 추천드립니다!

고생하셨어요 👍🏼

imjinbro commented on src/test/java/cooper/chess/piece/PawnTest.java:
```diff
@@ -1,22 +1,36 @@
 
     @Test
-    @DisplayName("흰색폰이 생성되어야 한다.")
+    public void create_defaultConstructor() {
```
junit5 > @DisplayName 을 사용하여 해당 테스트가 어떤 역할인지 표시하는 것도 해보시길 권해드립니다!

pbg0205 replied:
리뷰해주신대로 각각의 테스트에 @display 어노테이션을 추가해보았습니다!

imjinbro reviewed (changes requested):
안녕하세요 리뷰어 콜린입니다.
몇가지 코멘트를 남겨드렸습니다.
메소드를 분리하는 것과 pawnListMap 복잡한 구조를 수정 어떻게 할지 조금 더 고민해보시면 좋을듯 하여 Request changes 드립니다.

pbg0205 committed:
style : Board 상태 반환 메서드명 변경

- 이전의 출력 기능으로 혼동할 수 있는print 메서드명을 getBoardStatus
로 메서드명을 변경하였습니다.

- 각 행에 따른 상태를 반환하는 메서드 또한 getEachRowStatus로 메서드명
으로 변경하였습니다.

pbg0205 committed:
feat : chessGame 클래스 생성

- 체스게임이 작동하는 로직을 구분해서 관리하기 위해 ChessGame class를
추가하여 구현하였습니다.

- chessGame에서 메서드 분리를 체스판 세팅과 체스판 실행을 구분하기
위해 init 메서드와 play 메서드를 분리하였습니다.

pbg0205 committed:
test : 테스트 코드 @DisplayName 어노테이션 추가

- 해당 test 코드의 의미를 명확하게 확인하기 위해 테스트 코드 상단에
@DisplayName 어노테이션을 추가하였습니다.

pbg0205 committed:
feat : PawnGroup 클래스 추가

- Board 클래스 내부에 있던 pawnListMap을 별도의 클래스를 분리했습니다.
해당 클래스를 분리를 통해 Map의 불변성을 유지하고 기능에 제한을 두고자
하였습니다.

honux77 commented:
수고하셨어요!
